agent_controller/agent_controller.c

```diff
@@ -0,0 +1,869 @@
+
+  return agent;
+}
+
+/**
+ * @brief Build a JSON payload for updating agents.
+ *
+ * @param[in] agents List of agents to include in the payload.
+ * @param[in] update Optional update template to override agent fields.
+ *
+ * @return A newly allocated JSON string (unformatted) representing the update payload.
+ *         The caller is responsible for freeing the returned string using `g_free()`.
+ */
+static gchar *
+agent_controller_build_patch_payload (agent_controller_agent_list_t agents,
+                                      agent_controller_agent_update_t update)
+{
+  if (!agents || agents->count <=0)
+    return NULL;
+
+  cJSON *patch_body = cJSON_CreateObject ();
+
+  for (int i = 0; i < agents->count; ++i)
+    {
+      agent_controller_agent_t agent = agents->agents[i];
+      if (!agent || !agent->agent_id)
+        continue;
+
+      cJSON *agent_obj = cJSON_CreateObject ();
+
+      // authorized
+      int use_authorized = agent->authorized;
+      if (update && update->authorized != -1)
+        use_authorized = update->authorized;
+      cJSON_AddBoolToObject(agent_obj, "authorized", use_authorized);
+
+      // min_interval
+      int use_min_interval = agent->min_interval;
+      if (update && update->min_interval != -1)
+        use_min_interval = update->min_interval;
+      cJSON_AddNumberToObject(agent_obj, "min_interval", use_min_interval);
+
+      // heartbeat_interval
+      int use_heartbeat_interval = agent->heartbeat_interval;
+      if (update && update->heartbeat_interval != -1)
+        use_heartbeat_interval = update->heartbeat_interval;
+      cJSON_AddNumberToObject(agent_obj, "heartbeat_interval", use_heartbeat_interval);
+
+      // Config block
+      cJSON *config_obj = NULL;
+
+      // Use schedule override if given, else agent's own
+      agent_controller_config_schedule_t schedule_to_use =
+        (update && update->schedule_config) ? update->schedule_config
+                                            : agent->schedule_config;
+      if (schedule_to_use && schedule_to_use->schedule)
+        {
+          config_obj = cJSON_CreateObject ();
+
+          cJSON *schedule_obj = cJSON_CreateObject ();
+          cJSON_AddStringToObject (schedule_obj, "schedule",
+                                   schedule_to_use->schedule);
+          cJSON_AddItemToObject (config_obj, "schedule", schedule_obj);
+        }
+
+      if (config_obj)
+        {
+          cJSON_AddItemToObject (agent_obj, "config", config_obj);
+        }
+
+      cJSON_AddItemToObject (patch_body, agent->agent_id, agent_obj);
+    }
+
+  gchar *payload = cJSON_PrintUnformatted (patch_body);
+  cJSON_Delete (patch_body);
+
+  return payload;
+}
+
+/**
+ * @brief Creates a new Agent Controller connector.
+ */
+agent_controller_connector_t
+agent_controller_connector_new (void)
+{
+  agent_controller_connector_t connector;
+
+  connector = g_malloc0 (sizeof (struct agent_controller_connector));
+  if (!connector)
+    return NULL;
```

Comment:
```suggestion
```
Glib guarantees that memory is returned.
[https://docs.gtk.org/glib/func.malloc0.html](https://docs.gtk.org/glib/func.malloc0.html)
"If the allocation fails (because the system is out of memory), the program is terminated."